### PR TITLE
port kvstore from nim-beacon-chain

### PR DESCRIPTION
**Pull request by arnetheduck**
* remove old database backends
* use kvstore in trie database

### CODE REVIEW

File: eth/db/kvstore.nim [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+    ## Key-Value store virtual interface
+    obj: RootRef
+    putProc: PutProc
+    getProc: GetProc
+    delProc: DelProc
+    containsProc: ContainsProc
+
+template put*(db: KvStoreRef, key, val: openArray[byte]): KvResult[void] =
+  ## Store ``value`` at ``key`` - overwrites existing value if already present
+  db.putProc(db.obj, key, val)
```
Comment (zah) [inline eth/db/kvstore.nim:42]: Double evaluation of the `db` parameter. You can go for the usual fix:

```nim
template put*(dbParam: KvStoreRef, key, val: openArray[byte]): KvResult[void] =
  ## Store ``value`` at ``key`` - overwrites existing value if already present
  let db = dbParam
  db.putProc(db.obj, key, val)
```

Comment (zah) [reply eth/db/kvstore.nim:42]: Ditto for the rest of the file.

File: eth/db/kvstore_sqlite3.nim [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+  ./kvstore
+
+export kvstore
+
+type
+  SqStoreRef* = ref object of RootObj
+    env: ptr sqlite3
+    selectStmt, insertStmt, deleteStmt: ptr sqlite3_stmt
+
+template checkErr(op, cleanup: untyped) =
```
Comment (zah) [inline eth/db/kvstore_sqlite3.nim:17]: How come you never use the `cleanup` parameter in this file? None of the sqlite calls below require cleanup?

File: eth/db/kvstore.nim [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+  KvStoreRef(
+    obj: x,
+    putProc: putImpl[T],
+    getProc: getImpl[T],
+    delProc: delImpl[T],
+    containsProc: containsImpl[T]
+  )
+
+proc get*(db: MemStoreRef, key: openArray[byte], onData: DataProc): KvResult[bool] =
+  let key = @key
```
Comment (zah) [inline eth/db/kvstore.nim:87]: This should get the same TODO comment as the two functions below.

File: eth/trie/hexary.nim [hunk 4]
```diff
@@ -564,7 +564,7 @@ proc mergeAt(self: var HexaryTrie, orig: Rlp, origHash: KeccakHash,
   if orig.isEmpty:
     return origWithNewValue()
 
-  doAssert orig.isTrieBranch
+  doAssert orig.isTrieBranch, $orig
```
Comment (zah) [inline eth/trie/hexary.nim:567]: Is `doAssert` lazy with regard to the string parameter? `$(rlp)` is not cheap

Comment (zah) [reply eth/trie/hexary.nim:567]: I've just checked that it is, so it's OK

File: eth/db/kvstore_sqlite3.nim (hunk 2 above)
Comment (arnetheduck) [reply eth/db/kvstore_sqlite3.nim:17]: https://github.com/status-im/nim-eth/pull/230/files#diff-50ae87a573f6f1c02f8e6e155b691ba8R115
https://github.com/status-im/nim-eth/pull/230/files#diff-50ae87a573f6f1c02f8e6e155b691ba8R148

poor man's destructors I guess - but without move support and SqStore being `ref`..

File: eth/trie/hexary.nim (hunk 4 above)
Comment (arnetheduck) [reply eth/trie/hexary.nim:567]: yeah, I checked as well ;)

**Pull request merged**